src/builtin/builtin_source.c

```diff
@@ -16,27 +16,41 @@
 #include "utility.h"
 #include "shell.h"
 
-static void start_script(char **argv, env_t *env, int len_argv)
+char **get_arr_from_fd(int fd, char *filepath, char *str)
 {
-    char *str = NULL;
-    char **arr = NULL;
-    int fd = open(argv[1], O_RDONLY);
-    bool should_close = false;
     struct stat st_buff;
+    char **arr = NULL;
 
-    if (fd == -1)
-        return;
-    stat(argv[1], &st_buff);
+    stat(filepath, &st_buff);
```

Comment:
You do not check if read or stat returned an error so if we give you an invalid file the behavior is undefined and could make ash crash.

Reply:
Ok I lied since you open the file beforehand.